machine-check/src/swarm.rs

```diff
@@ -0,0 +1,715 @@
+                    let later_log = log_filter(log, sub(later));
+                    let extra = later_log & !filtered.clone();
+                    if extra.first_one().is_some() {
+                        errors.push(Error::LaterInvolvedRoleMoreSubscribed {
+                            edge: edge.id(),
+                            later: later.clone(),
+                            active: active.clone(),
+                            events: extra.iter_ones().map(|i| log[i].clone()).collect(),
+                        });
+                    }
+                }
+            }
+
+            // choice determinacy
+            let first_event = &log[0];
+            for later in &graph[target].roles {
+                if !sub(later).contains(first_event) {
+                    errors.push(Error::LaterInvolvedNotGuarded(edge.id(), later.clone()));
+                }
+            }
+        }
+    }
+    errors
+}
+
+pub fn from_json(
+    proto: SwarmProtocol,
+    subs: &Subscriptions,
+) -> (super::Graph, Option<NodeId>, Vec<String>) {
+    let (g, i, e) = prepare_graph(proto, subs);
+    (to_swarm(&g), i, e.map(Error::convert(&g)))
+}
+
+fn prepare_graph(
+    proto: SwarmProtocol,
+    subs: &Subscriptions,
+) -> (Graph, Option<NodeId>, Vec<Error>) {
+    let mut errors = Vec::new();
+    let mut graph = Graph::new();
+    let mut nodes = HashMap::new();
+    for t in proto.transitions {
+        tracing::debug!("adding {} --({:?})--> {}", t.source, t.label, t.target);
+        let source = *nodes
+            .entry(t.source.clone())
+            .or_insert_with(|| graph.add_node(Node::new(t.source)));
+        let target = *nodes
+            .entry(t.target.clone())
+            .or_insert_with(|| graph.add_node(Node::new(t.target)));
+        let edge = graph.add_edge(source, target, t.label.clone());
+        if t.label.log_type.len() == 0 {
+            errors.push(Error::LogTypeEmpty(edge));
+        }
+        tracing::debug!("added {:?} --> {:?}", source, target);
+    }
+    let initial = if let Some(idx) = nodes.get(&proto.initial) {
+        tracing::debug!("initial state {:?}", idx);
+        *idx
+    } else {
+        errors.push(Error::InitialStateDisconnected);
+        return (graph, None, errors);
+    };
+    let no_empty_logs = errors.is_empty();
+
+    // compute the needed Node information
+    // - first post-order walk to propagate non-loop roles back
+    // - then keep fixing loop-ends until graph is stable
+    let mut walk = DfsPostOrder::new(&graph, initial);
+    let mut loop_end = HashSet::new();
+    let mut guards = HashSet::new();
+    let mut events = HashMap::new();
+    while let Some(node_id) = walk.next(&graph) {
+        let active = active(&graph, node_id);
+        graph[node_id].active = active;
+        let roles = involved(&graph, node_id, &subs, &mut loop_end);
+        // FIXME: transition into final state without any subscription will make final state a loop end!
```

Comment:
I’m currently too sleepy to understand the first part, but the second part is the same to my loop_end loop, just less efficient because it checks many nodes that don’t need to be checked.

Reply:
An additional item that would optimize it greatly:
// changed_nodes: Set<Node> = all_nodes_in_postorder